Simulator/Utils/Factory.h

```diff
@@ -0,0 +1,102 @@
+/**
+ * @file Factory.h
+ * 
+ * @ingroup Simulator/Utils
+ *
+ * @brief A factory template for creating factory class for any subsystem that requires a 
+ *       factory to create a (singleton) concrete class instance at runtime, based on a class name specified in the configuration file.
+ *       
+ */
+
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
+
+/**
+ * The templated CreateFunctionMap allows each class to create a concrete instance of itself at runtime. 
+ * This approach helps to keep the code organized and avoids circular dependencies that can make the code less error-prone.
+ * 
+ * How it avoids circular dependency?
+ * If a map was passed from the base class with all the instance of its derived class, it would create a circular dependency. 
+ * Meaning, that the derived class would need to reference the base class,  
+ * and the base class would also need to reference the derived class for the sake of this map, leading to an infinite loop of references.
+ * 
+ * In addition, the templated CreateFunctionMap safeguards against creating a Factory code of an undesired type.
+ * A static assert is used to check for any undesired type; thereby generating a compile-time error for any request to instantiate 
+ * a template for a type that has not been explicitly defined.
+ * 
+*/
+
+template <typename T> std::map<std::string, T *(*)(void)> CreateFunctionMap(T *placeholder);
```

Comment:
One question: why is this a function and not a static method of the template class?

Reply:
Good suggestion, will make this a static method of Factory template.

Reply:
If you do that (and it works), then is it necessary to pass in a parameter at all? The template `T` will already be bound within the method to an actual base class, so I think you can just test it.

Reply:
It works! I really appreciate you guiding me through this. I learned a lot.